packages/blocks/src/code-block/components/code-actions.ts

```diff
@@ -0,0 +1,217 @@
+    return this.anchor.model;
+  }
+
+  private _copyCode = () => {
+    const anchor = this.anchor;
+    const slice = Slice.fromModels(this.model.doc, [this.model]);
+    anchor.std.clipboard
+      .copySlice(slice)
+      .then(() => {
+        toast(anchor.host, 'Copied to clipboard');
+      })
+      .catch(e => {
+        toast(anchor.host, 'Copied failed, something went wrong');
+        console.error(e);
+      });
+  };
+
+  private _duplicateCode = () => {
+    const { text, language, wrap, flavour } = this.model;
+    // FIXME
+    const newProps = { text: text.clone(), language, wrap, flavour };
+    const [codeId] = this.model.doc.addSiblingBlocks(this.model, [newProps]);
+    const editorHost = this.anchor.host;
+
+    editorHost.selection.setGroup('note', [
+      editorHost.selection.create('block', {
+        blockId: codeId,
+      }),
+    ]);
+  };
+
+  private _popMoreActions = (e: MouseEvent) => {
+    if (this.anchor.readonly) return;
+
+    this.style.visibility = 'visible';
+    const el = e.currentTarget as HTMLElement;
+
+    const items: Menu[] = [
+      {
+        type: 'action',
+        name: this.model.wrap ? 'Cancel Wrap' : 'Wrap',
+        icon: this.model.wrap ? CancelWrapIcon : WrapIcon,
+        select: () => {
+          this.anchor.setWrap(!this.model.wrap);
+        },
+      },
+
+      {
+        type: 'action',
+        name: 'Duplicate',
+        select: () => {
+          this._duplicateCode();
+        },
+        icon: DuplicateIcon,
+      },
+
+      {
+        type: 'group',
+        name: '',
+        children: () => [
+          {
+            type: 'action',
+            class: 'delete-item',
+            name: 'Delete',
+            select: () => {
+              this.model.doc.deleteBlock(this.model);
+            },
+            icon: DeleteIcon,
+          },
+        ],
+      },
+    ];
+
+    popMenu(el, {
+      placement: 'bottom-end',
+      middleware: [
+        offset(5),
+        autoPlacement({
+          allowedPlacements: ['bottom-start', 'bottom-end'],
+        }),
+      ],
+      options: {
+        items,
+        onClose: () => {
+          this.style.visibility = '';
+        },
+      },
+    });
+  };
+
+  override render() {
+    return html` <div contenteditable="false" class="code-actions">
+      <icon-button
+        class="code-action-button ai"
+        data-testid="ask-ai-button"
+        width="auto"
+        height="24px"
+        ?disabled=${this.anchor.readonly}
+      >
+        ${AIStarIcon} Ask AI
+      </icon-button>
+
+      <icon-button
+        class="code-action-button"
+        data-testid="copy-button"
+        width="auto"
+        height="24px"
+        @click=${this._copyCode}
+      >
+        ${CopyIcon}
+
+        <affine-tooltip tip-position="top" .offset=${5}
+          >Copy Code</affine-tooltip
+        >
+      </icon-button>
+
+      <!-- Not implemented -->
+      <icon-button
+        class="code-action-button hidden"
```

Comment:
We don't support the comment function yet. The design team should have updated the design. The comment button is currently not needed.